File: bae-core/src/library/download_snapshot.rs
//! The in-memory download (pin) queue snapshot. Single source of truth for the
//! Storage Manager's Downloads pane and the per-release "Downloading..." badge
//! when a pin is queued or in flight.

use super::release_queue::{
    build_release_queue_snapshot, ReleaseQueueOp, ReleaseQueueProgress, ReleaseQueueSnapshot,
    ReleaseQueueState,
};

use super::LibraryError;

#[derive(Debug, Clone, PartialEq)]
pub struct DownloadTransferProgress {
    pub bytes_done: u64,
    pub bytes_total: u64,
    pub fraction: f64,
}

impl DownloadTransferProgress {
    pub fn new(release_id: &str, bytes_done: u64, bytes_total: u64) -> Result<Self, LibraryError> {
        if bytes_total == 0 {
            return Err(LibraryError::Storage(format!(
                "pin release {release_id}: byte total is zero"
            )));
        }
        if bytes_done > bytes_total {
            return Err(LibraryError::Storage(format!(
                "pin release {release_id}: bytes done exceeds byte total"
            )));
        }
        if bytes_total > i64::MAX as u64 {
            return Err(LibraryError::Storage(format!(
                "pin release {release_id}: byte total exceeds display range"
            )));
        }
        Ok(Self {
            bytes_done,
            bytes_total,
            fraction: bytes_done as f64 / bytes_total as f64,
        })
    }
}

pub type DownloadState = ReleaseQueueState<DownloadTransferProgress>;
pub type DownloadOp = ReleaseQueueOp<(), DownloadTransferProgress>;
pub type DownloadProgress = ReleaseQueueProgress;
pub type DownloadSnapshot = ReleaseQueueSnapshot<(), DownloadTransferProgress>;

pub fn build_download_snapshot(downloads: &[DownloadOp], paused: bool) -> DownloadSnapshot {
    build_release_queue_snapshot(downloads, paused)
}

File: bae-windows/DownloadSnapshot.cs
using System.Collections.Generic;

namespace Bae.Windows;

/// <summary>The in-memory download (pin) queue snapshot. Drives the Downloads
/// pane, and the storage row reads it to tell a pinning release (queued or
/// downloading) from an idle one so the row can offer to cancel the pin.</summary>
public sealed class DownloadSnapshot
{
    public List<DownloadOp> Downloads { get; set; } = new();

    /// <summary>Per-state counts rolled up across the queue — drives the pane
    /// header and the retry gate.</summary>
    public DownloadProgress Total { get; set; } = new();

    /// <summary>True when the user paused the queue — drives the pause/resume
    /// toggle.</summary>
    public bool Paused { get; set; }
}

/// <summary>Per-state counts for the download queue.</summary>
public sealed class DownloadProgress
{
    public uint Queued { get; set; }
    public uint Active { get; set; }
    public uint Failed { get; set; }
}

/// <summary>One queued download — a whole release being pinned.</summary>
public sealed class DownloadOp
{
    public string ReleaseId { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;
    public long FileCount { get; set; }
    public long TotalSize { get; set; }

    /// <summary>"queued", "active", or "failed".</summary>
    public string State { get; set; } = string.Empty;

    /// <summary>The active transfer's byte/file progress when <see cref="State"/>
    /// is "active".</summary>
    public DownloadTransferProgress? Progress { get; set; }

    /// <summary>The failure message when <see cref="State"/> is "failed".</summary>
    public string? Error { get; set; }
}

/// <summary>Byte and file progress for the active download.</summary>
public sealed class DownloadTransferProgress
{
    public ulong BytesDone { get; set; }
    public ulong BytesTotal { get; set; }
    public double Fraction { get; set; }
}
